feat: lint against the tactic.skipAssignedInstances option in mathlib

grunweg opened the pull request:
This means we can remove the corresponding technical debt entry. We also correct some outdated details in the Linter/Style.lean docstring.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
feat: lint against the tactic.skipAssignedInstances option in mathlib

This means we can remove the corresponding technical debt entry.

grunweg commented:
@adomani This runs into another instance of the "`set_option name in` is not recognised", which is somewhat unfortunate.
Do you know how to this fix? (I think we came across this for sure, but don't remember what the conclusion was.)

adomani commented:
Michael, did you fix the error since you posted the message?  I do not see any issues at the moment.

grunweg commented:
Damiano: the issue is that tests pass :-) More specifically, it's in these lines (in `MathlibTest/LintStyle.lean`):
```
-- TODO: this should warn!
set_option tactic.skipAssignedInstances true in
lemma skipAssInst'' : True := trivial
```
the fact that *no* warning is emitted.

adomani commented on MathlibTest/LintStyle.lean:
```diff
@@ -117,7 +117,28 @@ lemma tactic4 : True := by
+-/
+#guard_msgs in
+set_option tactic.skipAssignedInstances false
+lemma skipAssInst' : True := trivial
```
```suggestion
```
Does this line do anything at all?

adomani commented:
Ah, that is because you should match also on `set_option ... in`.  Let me look more in depth at the code and how to catch that.

adomani commented:
I guess that, for this linter, it is important to get access to all `set_option ... in`, and depending on the option, to decide if it should emit a warning or not: some `set_option`s should only exist scoped, others should not exist at all.

adomani commented:
Michael, given that there are quite a few updates, fixes and renames that this PR is doing, on top of adding the new linting, would you mind splitting the PR into the "house-keeping" PR with updates and renames, and then a dependent PR with the actual changes to the linting?

I can approve the housekeeping PR immediately, and then we can focus on the best approach to linting `set_option in`.

grunweg commented:
I'm not sure if 8 lines of diff are worth a new PR, but here you go: #20879

grunweg committed:
fix: remove outer withSetOptionIn; add tests for set_option ... in

The outermost set_option is linted against, inner set_options are not.
At least, it's an improvement...

grunweg commented:
Removing the `withSetOptionIn` already helped: now outer `set_option ... in` are linted.
This doesn't work for *nested* set_option ... in, though (see the new tests).

adomani commented:
> Removing the `withSetOptionIn` already helped: now outer `set_option ... in` are linted. This doesn't work for _nested_ set_option ... in, though (see the new tests).

I presume that this is because the filter on the syntax is `find?`, which returns the first instance that it finds, rather than all occurrences.

adomani commented:
I do not know if this is enough of a performance issue to require a careful scanning, but here is a proposal.

1. If the `Syntax` does not start with `set_option ... in` and is not `set_option ...`, do nothing.
2. If it is `set_option` or `set_option ... in`, then scan all syntax, extracting the array of options and then continue with roughly the same logic as now, except that the script should take care of which `in`s are allowed and which ones are not.

adomani commented:
Wow, the "new" linter found two exceptions!!

grunweg committed:
chore: remove debugging options

grunweg commented:
Fixed (these were pre-existing, but linting set_option in caught it :tada:) and rebased.
